csi/src/controller.rs

```diff
@@ -362,189 +378,195 @@ impl ControllerImpl {
+            if vol_id.is_empty() {
+                return Err((
+                    RpcStatusCode::INVALID_ARGUMENT,
+                    anyhow!("volume ID missing in request"),
+                ));
+            }
 
-        // Do not return gRPC error when delete failed for idempotency
-        let vol_res = self.meta_data.get_volume_by_id(vol_id);
-        if let Some(vol) = vol_res {
-            let node_res = self.meta_data.get_node_by_id(&vol.node_id);
-            if let Some(node) = node_res {
-                let client = MetaData::build_worker_client(&node);
-                let worker_delete_res = client.worker_delete_volume(&req).context(format!(
-                    "failed to delete volume ID={} on node ID={}",
-                    vol_id, vol.node_id,
+            let rpc_type = ControllerServiceCapability_RPC_Type::CREATE_DELETE_VOLUME;
+            if !self_inner.validate_request_capability(rpc_type) {
+                return Err((
+                    RpcStatusCode::INVALID_ARGUMENT,
+                    anyhow!(format!("unsupported capability {:?}", rpc_type)),
                 ));
-                match worker_delete_res {
-                    Ok(_) => info!("successfully deleted volume ID={}", vol_id),
-                    Err(e) => {
-                        warn!(
-                            "failed to delete volume ID={} on node ID={}, the error is: {}",
-                            vol_id,
-                            vol.node_id,
-                            util::format_anyhow_error(&e),
-                        );
+            }
+
+            // Do not return gRPC error when delete failed for idempotency
+            let vol_res = self_inner.meta_data.get_volume_by_id(vol_id).await;
+            if let Some(vol) = vol_res {
+                let node_res = self_inner.meta_data.get_node_by_id(&vol.node_id).await;
+                if let Some(node) = node_res {
+                    let client = MetaData::build_worker_client(&node);
+                    let req_clone = req.clone();
+                    let worker_delete_res =
+                        smol::unblock(move || client.worker_delete_volume(&req_clone))
+                            .await
+                            .context(format!(
+                                "failed to delete volume ID={} on node ID={}",
+                                vol_id, vol.node_id,
+                            ));
+                    match worker_delete_res {
+                        Ok(_) => info!("successfully deleted volume ID={}", vol_id),
+                        Err(e) => {
+                            warn!(
+                                "failed to delete volume ID={} on node ID={}, the error is: {}",
+                                vol_id,
+                                vol.node_id,
+                                util::format_anyhow_error(&e),
+                            );
+                        }
                     }
```

Comment:
Why not call `client.xxx_async` directly?

I'm not familiar with `grpcio`. The code below just compiles. What's the difference with `unblock`?

```rust
                    let client = MetaData::build_worker_client(&node);
                    let worker_delete_res =
                        util::wait_ok(client.worker_delete_volume_async(&req)).await;

                    match worker_delete_res {
                        Ok(_) => info!("successfully deleted volume ID={}", vol_id),
                        Err(e) => {
                            warn!(
                                "failed to delete volume ID={} on node ID={}, the error is: {}",
                                vol_id,
                                vol.node_id,
                                util::format_anyhow_error(&anyhow::Error::new(e)),
                            );
                        }
                    }
```

```rust
/// Wait for `Result<F, E>` where `F: Future<Output = T>`, then return `Result<T, E>`
#[allow(clippy::future_not_send)]
pub async fn wait_ok<F, T, E>(res: Result<F, E>) -> Result<T, E>
where
    F: Future<Output = T>,
{
    match res {
        Ok(f) => Ok(f.await),
        Err(e) => Err(e),
    }
}
```

Reply:
Thanks. I have noticed that grpcio provides async client calls.